parser/util/Environment.js

```diff
@@ -0,0 +1,78 @@
+import { LinkedList } from './LinkedList';
+import { _ } from 'underscore';
+
+/**
+ * A special key-value pair linked list that can lookup values of a certain key.
+ */
+export class Environment extends LinkedList {
+  /**
+   * Creates a new environment with an associated key-value pair. The key nor
+   * value cannot be undefined.
+   * @param {*} key   the searchable identifier that maps to the value
+   * @param {*} value the value associated with the key
+   * @throws {Error} when key or value is undefined
+   * @return a new Environment with the appended key-value pair
+   */
+  push (key, value) {
```

Comment:
Since `push` no longer supports single-parameter calls, hence an interface of the superclass is not supported by the subclass, I don't see any more reason to keep `Environment` as a subclass for `LinkedList`. This violates Liskov's substitution principle.

While this does not matter very much in a dynamic language like JS, I think we should change `Environment` to be its own independent class as good practice, and so that people who will use this for TS projects will not make the mistake of assigning an Environment instance to a LinkedList variable and calling a single-parameter `push`.

What do you think?

Reply:
Your reasoning makes sense to me. Will do isolate Environment as its own class. Will add that in the next commit.
However, I would like to keep the persistent LinkedList since will be using that to list errors/warnings, if that's okay :)

Reply:
Sure